content/en/blog/_posts/2024-04-xx-structured-authz-beta.md

```diff
@@ -0,0 +1,328 @@
+      # only send resource requests to the webhook
+      - expression: has(request.resourceAttributes)
+      # only intercept requests for CRDs
+      - expression: request.resourceAttributes.resource.resource = "customresourcedefinitions"
+      - expression: request.resourceAttributes.resource.group = ""
+      # only intercept update, patch, delete, or deletecollection requests
+      - expression: request.resourceAttributes.verb in ['update', 'patch', 'delete','deletecollection']
+  - type: Node
+  - type: RBAC
+```
+
+### Preventing unnecessarily nested webhooks
+A system administrator wants to apply specific validations to requests before
+handling them off to webhooks using frameworks like Open Policy Agent. In the
+past, this would require running nested webhooks within the one added to the
+authorization chain to achieve the desired result. The Structured Authorization
+Configuration feature simplifies this process, offering a structured API to
+selectively trigger additional webhooks when needed. It also enables
+administrators to set distinct failure policies for each webhook, ensuring more
+consistent and predictable responses.
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1beta1
+kind: AuthorizationConfiguration
+authorizers:
+  - type: Webhook
+    name: system-crd-protector
+    webhook:
+      unauthorizedTTL: 30s
+      timeout: 3s
+      subjectAccessReviewVersion: v1
+      matchConditionSubjectAccessReviewVersion: v1
+      failurePolicy: Deny
+      connectionInfo:
+        type: KubeConfig
+        kubeConfigFile: /files/kube-system-authz-webhook.yaml
+      matchConditions:
+      # only send resource requests to the webhook
+      - expression: has(request.resourceAttributes)
+      # only intercept requests for CRDs
+      - expression: request.resourceAttributes.resource.resource = "customresourcedefinitions"
+      - expression: request.resourceAttributes.resource.group = ""
+      # only intercept update, patch, delete, or deletecollection requests
+      - expression: request.resourceAttributes.verb in ['update', 'patch', 'delete','deletecollection']
+  - type: Node
+  - type: RBAC
+  - name: opa
+    type: Webhook
+    webhook:
+      unauthorizedTTL: 30s
+      timeout: 3s
+      subjectAccessReviewVersion: v1
+      matchConditionSubjectAccessReviewVersion: v1
+      failurePolicy: Deny
+      connectionInfo:
+        type: KubeConfig
+        kubeConfigFile: /files/opa-default-authz-webhook.yaml
+      matchConditions:
+      # only send resource requests to the webhook
+      - expression: has(request.resourceAttributes)
+      # only intercept requests to default namespace
+      - expression: request.resourceAttributes.namespace == 'default'
+      # don't intercept requests from default service accounts
+      - expression: !('system:serviceaccounts:default' in request.user.groups)
+```
+
+## What's next?
+For Kubernetes v1.31, we expect the feature to stay in beta while we get more
```

Comment:
> For Kubernetes v1.31, we expect the feature to stay in beta while we get more feedback.

I feel like this sentence is missing something -- is this only the process for v1.31? Is this is a general beta feature process? Who is providing the feedback?

Reply:
will update to
> From Kubernetes 1.30, the feature is in beta and enabled by default. For Kubernetes v1.31, we expect the feature to stay in beta while we get more feedback from users.

In general, we keep beta features around for one or two releases before promoting it to GA.